pkg/controller/useraccount/useraccount_controller_test.go

```diff
@@ -1,134 +1,163 @@
+	r := &ReconcileUserAccount{
+		client: client,
+		scheme: s,
+	}
 
-func TestEnsureIdentity(t *testing.T) {
-	logf.SetLogger(logf.ZapLogger(true))
-	username := "johnsmith"
-	userID := "34113f6a-2e0d-11e9-be9a-525400fb443d"
-	s := scheme.Scheme
-	err := apis.AddToScheme(s)
-	require.NoError(t, err)
+	req := newReconcileRequest(username)
+	identity := &userv1.Identity{}
+	user := &userv1.User{}
 
-	t.Run("create_identity", func(t *testing.T) {
-		userAcc := newUserAccount(username, userID)
-		client := fake.NewFakeClient()
-		reconciler := &ReconcileUserAccount{
+	t.Run("deleted_account_ignored", func(t *testing.T) {
+		// given
+		// No user account exists
+		objs := []runtime.Object{}
+		client := fake.NewFakeClient(objs...)
+		r := &ReconcileUserAccount{
 			client: client,
 			scheme: s,
 		}
-
-		newIdentity, created, err := reconciler.ensureIdentity(userAcc)
-		assert.NoError(t, err)
-		assert.True(t, created)
-		assert.NotNil(t, newIdentity)
+		//when
+		res, err := r.Reconcile(req)
+
+		//then
+		require.NoError(t, err)
+		assert.Equal(t, reconcile.Result{}, res)
+
+		// Check the user is not created
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user)
+		require.Error(t, err)
+		assert.True(t, errors.IsNotFound(err))
+
+		// Check the identity is not created
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: getIdentityName(userAcc)}, identity)
+		require.Error(t, err)
+		assert.True(t, errors.IsNotFound(err))
 	})
 
-	t.Run("get_identity", func(t *testing.T) {
-		userAcc := newUserAccount(username, userID)
-		identity := newIdentity(userAcc)
-		client := fake.NewFakeClient(identity)
-		reconciler := &ReconcileUserAccount{
-			client: client,
-			scheme: s,
-		}
+	t.Run("create_user", func(t *testing.T) {
+		//when
+		res, err := r.Reconcile(req)
+
+		//then
+		require.NoError(t, err)
+		assert.Equal(t, reconcile.Result{}, res)
+
+		// Check that the user account status has been updated
+		updatedAcc := &toolchainv1alpha1.UserAccount{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: req.Namespace, Name: userAcc.Name}, updatedAcc)
+		require.NoError(t, err)
+		assert.Equal(t, toolchainv1alpha1.StatusProvisioning, updatedAcc.Status.Status)
+
+		// Check the created user
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user)
+		require.NoError(t, err)
+		assert.Equal(t, userAcc.Name, user.Name)
+		assert.Len(t, user.GetOwnerReferences(), 1)
+		assert.Equal(t, updatedAcc.UID, user.GetOwnerReferences()[0].UID)
+
+		// Check the identity is not created yet
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: getIdentityName(userAcc)}, identity)
+		require.Error(t, err)
+		assert.True(t, errors.IsNotFound(err))
+	})
 
-		newIdentity, created, err := reconciler.ensureIdentity(userAcc)
-		assert.NoError(t, err)
-		assert.False(t, created)
-		assert.NotNil(t, newIdentity)
+	t.Run("create_identity", func(t *testing.T) {
```

Comment:
this subtest setup (and also the following ones) is a bit problematic because it relies on the previous subtest - it expects that the User already exists, so it's no possible to run it separately.

Reply:
That's correct. These subtests are different steps of the same test rather than independent tests. I still prefer to keep them as is. Otherwise the test will have unnecessarily duplicated code. I can add comments though...

Reply:
Added comments in https://github.com/codeready-toolchain/member-operator/pull/25/commits/81f9d57c0d73f11d0430e1ae57b28891038d1a96

Reply:
But then we are losing the big benefit of subtests - to be able to run them separately and when one subtest fails, then the other dependent ones should not be affected.
I'm not a fan of the (sub)tests that depend on previous ones (if not really needed as can be for example in UI tests, but this is not the case).  
As for your concern about duplicated code - I don't think that there would be too much of duplication. I tried to quickly rewrite the code - see here https://gist.github.com/MatousJobanek/4816087ca5f69df0434290086f022d14 could be still improved

Reply:
IMO it's really not a big deal in this case. But anyway.. here we go - https://github.com/codeready-toolchain/member-operator/pull/25/commits/358de797d11bab64e776286f73433ceba39ff619